include/pybind11/detail/eigen_tensor.h

```diff
@@ -0,0 +1,401 @@
+    }
+
+    static handle cast(Eigen::TensorMap<Type> &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast(&src, policy, parent);
+    }
+
+    static handle cast(Eigen::TensorMap<Type> *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    static handle
+    cast(const Eigen::TensorMap<Type> *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    template <typename C>
+    static handle cast_impl(C *src, return_value_policy policy, handle parent) {
+        object parent_object;
+        constexpr bool writeable = !std::is_const<C>::value;
+        switch (policy) {
+            case return_value_policy::reference:
+                parent_object = none();
+                break;
+
+            case return_value_policy::reference_internal:
+                // Default should do the right thing
+                if (!parent) {
+                    pybind11_fail("Cannot use reference internal when there is no parent");
+                }
+                parent_object = reinterpret_borrow<object>(parent);
+                break;
+
+            default:
+                // move, take_ownership don't make any sense for a ref/map:
+                pybind11_fail("Invalid return_value_policy for Eigen Map type, must be either "
+                              "reference or reference_internal");
+        }
+
+        object result = array_t<typename Type::Scalar, compute_array_flag_from_tensor<Type>()>(
+            H::get_shape(*src), src->data(), parent_object);
+
+        if (!writeable) {
+            array_proxy(result.ptr())->flags &= ~detail::npy_api::NPY_ARRAY_WRITEABLE_;
+        }
+
+        return result.release();
+    }
+
+protected:
+    // TODO: Move to std::optional once std::optional has more support
+    std::unique_ptr<Eigen::TensorMap<Type>> value;
```

Comment:
Interesting.

Could this be useful also for `type_caster<Eigen::Tensor>` above? — I don't fully understand how all the parts work together here and above. What's on my mind: could that help avoid copies above? — Evidently, from this caster, it doesn't take a lot to avoid the `PYBIND11_TYPE_CASTER` macro.

Another loosely related question: what happens for `std::unique_ptr` conversions from C++ to Python, and `std::shared_ptr` both ways? Is the behavior well defined in all situations (compiler errors, clean runtime errors)?

I wouldn't want to expand the scope of this PR too much, just enough to be sure we don't set up undefined behavior traps. If there are traps and we cannot easily avoid them, it's important to document that somewhere prominent.

Reply:
@rwgk The eigen_matrix.h casters do something similar currently, that's probably what this PR is copy code from.

Reply:
The purpose of the std::unique_ptr is because TensorMap doesn't have a default constructor. Added a comment to make that more clear. The std::unique_ptr is explicitly hidden from the rest of the machinery, it never gets converted to/from Python.

Eigen::Tensor does something similar with Eigen::Ref (although Eigen::Ref is more efficient and avoids dynamic allocation).